relays/finality-relay/src/finality_loop.rs

```diff
@@ -0,0 +1,558 @@
+				return Ok(Some((header, finality_proof)));
+			}
+			(true, None) => return Err(Error::MissingMandatoryFinalityProof(header.number())),
+			(false, Some(finality_proof)) => {
+				log::trace!(target: "bridge", "Header {:?} has persistent finality proof", header_number);
+				selected_finality_proof = Some((header, finality_proof));
+				prune_unjustified_headers::<P>(header_number, &mut unjustified_headers);
+			}
+			(false, None) => {
+				unjustified_headers.push(header);
+			}
+		}
+
+		header_number = header_number + One::one();
+	}
+
+	// see if we can improve finality by using recent finality proofs
+	if !unjustified_headers.is_empty() && !recent_finality_proofs.is_empty() {
+		const NOT_EMPTY_PROOF: &str = "we have checked that the vec is not empty; qed";
+
+		// we need proofs for headers in range unjustified_range_begin..=unjustified_range_end
+		let unjustified_range_begin = unjustified_headers.first().expect(NOT_EMPTY_PROOF).number();
+		let unjustified_range_end = unjustified_headers.last().expect(NOT_EMPTY_PROOF).number();
+
+		// we have proofs for headers in range buffered_range_begin..=buffered_range_end
+		let buffered_range_begin = recent_finality_proofs.first().expect(NOT_EMPTY_PROOF).0;
+		let buffered_range_end = recent_finality_proofs.last().expect(NOT_EMPTY_PROOF).0;
+
+		// we have two ranges => find intersection
+		let intersection_begin = std::cmp::max(unjustified_range_begin, buffered_range_begin);
+		let intersection_end = std::cmp::min(unjustified_range_end, buffered_range_end);
+		let intersection = intersection_begin..=intersection_end;
+
+		// find last proof from intersection
+		let selected_finality_proof_index = recent_finality_proofs
+			.binary_search_by_key(intersection.end(), |(number, _)| *number)
+			.unwrap_or_else(|index| index.saturating_sub(1));
+		let (selected_header_number, finality_proof) = &recent_finality_proofs[selected_finality_proof_index];
+		if intersection.contains(selected_header_number) {
+			// now remove all obsolete headers and extract selected header
+			let selected_header = prune_unjustified_headers::<P>(*selected_header_number, &mut unjustified_headers)
+				.expect("unjustified_headers contain all headers from intersection; qed");
+			selected_finality_proof = Some((selected_header, finality_proof.clone()));
+		}
+	}
+
+	// read all proofs from the stream, probably selecting updated proof that we're going to submit
+	loop {
+		let next_proof = finality_proofs_stream.next();
+		let finality_proof = match next_proof.now_or_never() {
+			Some(Some(finality_proof)) => finality_proof,
+			_ => break,
+		};
+		let finality_proof_target_header_number = match finality_proof.target_header_number() {
+			Some(target_header_number) => target_header_number,
+			None => {
+				continue;
+			}
+		};
+
+		let justified_header =
+			prune_unjustified_headers::<P>(finality_proof_target_header_number, &mut unjustified_headers);
```

Comment:
Why are we doing pruning in a loop? Wouldn't it be better to first run the loop and select the "best" finality proof to submit and only then prune everything once?

Reply:
Yes, it may be done - I'll probably do that in next PR. But please keep in mind that normally we'll either never reach this point with more than 1-2 headers in `unfinalized_headers`, or will reach it only once in a relayer lifetime. It is when relayer just have started and there are many unsynced headers from the last GRANDPA session.